### PR TITLE
[Coupons] M1 analytics

### LINKED ISSUE #6323: [Coupons]: Add tracks for Milestone 1
<tbody class="d-block" style="box-sizing: border-box; display: block !important; color: rgb(36, 41, 47); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><tr class="d-block" style="box-sizing: border-box; display: block !important;"><td class="d-block comment-body markdown-body  js-comment-body overflow-auto" style="box-sizing: border-box; padding: 16px; overflow: visible; display: block !important; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; line-height: 1.5; overflow-wrap: break-word; width: 838px;"><p dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px;"><br class="Apple-interchange-newline">For M1, we need a few Tracks events on the coupon list and details screen:</p>

Event Name | Trigger | Properties
-- | -- | --
*_coupons_loaded | Request to load coupons succeeded. | is_loading_more: true \| false
*_coupons_load_failed | Requested to load coupons failed. | error_context: $initiated-classnameerror_type: $errorTypeerror_description: $errorDescription
*_coupons_list_search_tapped | User tapped the search button on the coupon list. |  
*_coupon_details | User loaded/copied/shared the coupon code/tapped the delete button on the coupon details screen. | action: loaded \| copied_code \| shared_code \| tapped_delete

</td></tr></tbody><br class="Apple-interchange-newline">

### PR DESCRIPTION
**Pull request by 0nko**
Fixes #6323 and adds the missing event tracking for coupons.

**To test:**
1. Turn on Flight mode
2. Go to Coupons
3. Wait for a loading failure
4. Verify that `*_coupons_load_failed` is tracked
5. Leave coupons and turn off Flight mode
6. Go back to coupons
7. Verify that `*_coupons_loaded` is tracked
8. Tap on **Search** 
9. Verify that `*_coupons_list_search_tapped` is tracked
10. Go to a coupon detail
11. Verify that `*_coupon_details` with `loaded` action is tracked
12. Tap on the menu and select **Copy coupon code**
13. Verify that `*_coupon_details` with `copied_code` action is tracked
14. Tap on the menu and select **Share coupon**
15. Verify that `*_coupon_details` with `shared_code` action is tracked
16. Tap on the menu and select **Delete coupon**
17. Verify that `*_coupon_details` with `tapped_delete` action is tracked

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/details/CouponDetailsViewModel.kt [hunk 1]
```diff
@@ -52,6 +54,11 @@ class CouponDetailsViewModel @Inject constructor(
                     triggerEvent(Exit)
                 }
             }
+
+            AnalyticsTracker.track(
+                AnalyticsEvent.COUPON_DETAILS,
+                mapOf(AnalyticsTracker.KEY_COUPON_ACTION to AnalyticsTracker.KEY_COUPON_ACTION_LOADED)
+            )
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/details/CouponDetailsViewModel.kt]: NP, having this tracked after fetching the coupon will give a different result than iOS in some cases, for example if the user entered the screen, then left before the fetching finishes, in this case iOS will track this, while we won't.

WDYT about putting it outside the `launch` block to make it do the tracked initially? for all cases, we will be having a coupon cached in the DB, but if we are worried about this case, we can do something like this too to make sure it's tracked only when a coupon is loaded:
```kotlin
        coupon.filterNotNull()
            .take(1)
            .onEach {
                AnalyticsTracker.track(
                    AnalyticsEvent.COUPON_DETAILS,
                    mapOf(AnalyticsTracker.KEY_COUPON_ACTION to AnalyticsTracker.KEY_COUPON_ACTION_LOADED)
                )
            }
            .launchIn(viewModelScope)
```

**Review by hichamboushaba — APPROVED** [review]
Nice work @0nko, I just left a comment about a slightly different behavior than iOS, but it's not blocking.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/details/CouponDetailsViewModel.kt (hunk 1 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/details/CouponDetailsViewModel.kt]: I'm not too worried about it, I think putting it outside the `launch` block should be fine.

**Comment by 0nko** [comment]
Thanks @hichamboushaba, ready for another round.

**Review by hichamboushaba — APPROVED** [review]
Looks good, thanks for the changes @0nko.

There is a pending detekt issue, feel free to merge once it's fixed.

**Pull request merged**